pkg/metrics/collector/pod_ip_metrics.go

```diff
@@ -322,22 +337,92 @@ func (c *podIPMetricsCollector) setupDirectoryWatcher(dir string) error {
 	return nil
 }
 
+func (c *podIPMetricsCollector) countTotalIPs(subnet *net.IPNet) (uint64, error) {
+	ones, bits := subnet.Mask.Size()
+	if bits == 0 {
+		return 0, fmt.Errorf("invalid IP subnet mask %v", subnet.Mask)
+	}
+	if ones == bits {
+		ones = bits - 1
+	}
+	return uint64(1 << (bits - ones)), nil
+}
+
+func (c *podIPMetricsCollector) calculateAssignedIP(subnetFile string) error {
+	var assignedIPv4, assignedIPv6 uint64
+	networkConfigList, err := libcni.ConfListFromFile(subnetFile)
+	if err != nil {
+		glog.Errorf("Error reading network config from file %s: %v", subnetFile, err)
+		return err
+	}
+
+	var ipamConfig *allocator.IPAMConfig
+	for _, plugin := range networkConfigList.Plugins {
+		if plugin.Network.Type == "gke" {
+			ipamConfig, _, err = allocator.LoadIPAMConfig(plugin.Bytes, "")
+			if err != nil {
+				glog.Errorf("Error loading IPAM config: %v", err)
+				return err
+			}
+			break
+		}
+	}
+
+	for _, ranges := range ipamConfig.Ranges {
+		for _, subnets := range ranges {
+			convertedSubnet := &net.IPNet{
+				IP:   subnets.Subnet.IP,
+				Mask: subnets.Subnet.Mask,
+			}
+			if convertedSubnet.IP.To16() != nil && convertedSubnet.IP.To4() == nil {
+				totalIPv6, err := c.countTotalIPs(convertedSubnet)
```

Comment:
These two chunks look repetitive.. I don't have a good suggestion now though. Maybe we can look back to it until we handled `rangeStart` / `rangeEnd`?

Reply:
I extract the total IP calculation outside the loop, so there will be only one calculation chunk. I hope this will simplify the code.